File: epoch_conversion/ocaml_epoch_conversion/epoch_conversion.ml
(* comment *)
open Unix;;

let computeDateDoy year month day =
  let n1 = 275 * month / 9 in
  let n2 = (month + 9) / 12 in
  let n3 = (1 + ((year - 4 * (year / 4) + 2) / 3)) in
  let n = n1 - (n2 * n3) + day - 30 in
  n;;

let datetimeToEpoch years months day hours mins secs =
  let epoch_start = ref 1970 in
  let total_days = ref 0 in
  let total_secs = ref 0 in

  while  !epoch_start < years  do
    total_days := !total_days + (computeDateDoy !epoch_start 12 31);
    epoch_start := !epoch_start + 1
  done;

  total_days := !total_days + (computeDateDoy years months (day-1));
  total_secs := (!total_days * 86400) + (hours * 60 * 60) + (mins * 60) + secs;

  !total_secs;;

let main () =
  let tm = time() in
  let utc = gmtime tm in
  let year = 1900 + utc.tm_year in
  let month = 1 + utc.tm_mon in
  let day = utc.tm_mday in
  let hour = utc.tm_hour in
  let min = utc.tm_min in
  let sec = utc.tm_sec in
  if  Array.length Sys.argv <> 1 then
    begin
      Printf.printf "%s <noargs>\n" Sys.argv.(0);
      exit 1
    end
  else
    begin
    Printf.printf "epoch=<%d>\n" (datetimeToEpoch year month day hour min sec);
    exit 0
    end

let _ = main ()

File: epoch_conversion/c_epoch_conversion/epoch_conversion.c
#if defined(_WIN32)
  #define OS_WINDOWS 1 // Windows
#elif defined(_WIN64)
  #define OS_WINDOWS 1 // Windows
#else
  //error Unsupported operating system
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

void epochToDatetime( long );
int computeDateDoy( int, int, int );
long datetimeToEpoch( int y, int m, int mday, int h, int min, int s );

int main( int argc, char *argv[] ) {
  long epoch = 0L;
  time_t utc_datetime = 0;
  struct tm *utc = NULL;

  if( argc != 1 ) {
    fprintf(stderr, "Usage: %s <noargs>\n", argv[0]);
    exit(1);
  }

  utc_datetime = time(NULL);
  utc = gmtime(&utc_datetime);
  epoch = datetimeToEpoch(utc->tm_year + 1900, utc->tm_mon + 1, utc->tm_mday, utc->tm_hour, utc->tm_min, utc->tm_sec);
  printf("epoch=<%ld>\n", epoch);
  printf("utc=<%d/%d/%d %d:%d:%d>\n", utc->tm_year + 1900, utc->tm_mon + 1, utc->tm_mday, utc->tm_hour, utc->tm_min, utc->tm_sec);
  epochToDatetime(epoch);
  return 0;
}

// need to add logic to do the conversion
void epochToDatetime( long epoch ) {
  printf("epoch=<%ld>\n", epoch);
  printf("utc=<%s>\n", strtok(asctime(gmtime((time_t *) &epoch)), "\n"));
  printf("localtime=<%s>\n", strtok(asctime(localtime((time_t *) &epoch)), "\n"));
}

int computeDateDoy( int year, int month, int mday ) {
  int n1 = 275 * month / 9;
  int n2 = (month + 9) / 12;
  int n3 = (1 + ((year - 4 * (year / 4) + 2) / 3));
  int n = n1 - (n2 * n3) + mday - 30;

  return n;
}

long datetimeToEpoch( int years, int months, int mday, int hours, int mins, int secs ) {
  int epoch_start = 1970;
  int total_days = 0;
  int total_secs = 0;

  while( epoch_start < years ) {
    total_days = total_days + computeDateDoy(epoch_start, 12, 31);
    epoch_start = epoch_start + 1;
  }
  total_days = total_days + computeDateDoy(years, months, mday - 1);
  total_secs = (total_days * 86400) + (hours * 60 * 60) + (mins * 60) + secs;
  return total_secs;
}
